File: apps/web/src/types/analysis.ts
export interface PortfolioSnapshotRead {
  id: number;
  portfolio_id: number;
  created_at: string;
  total_value: number | null;
  num_positions: number;
  recent_earnings_coverage: number | null;
  avg_sentiment_score: number | null;
  risk_exposure_score: number | null;
  health_score: number | null;
  concentration_risk: number | null;
  daily_change: number | null;
  daily_change_pct: number | null;
}

export interface BenchmarkPoint {
  date: string;
  portfolio_pct: number;
  sp500_pct: number;
}

export interface PortfolioHistoryWithBenchmark {
  data: BenchmarkPoint[];
}

export interface SectorAllocation {
  sector: string;
  weight: number;
  value: number;
}

export interface EarningsInsights {
  holdings_with_recent_earnings: string[];
  positive_outlooks: string[];
  risk_warnings: string[];
  recommended_reviews: string[];
  sentiment_summary: Record<string, number>;
}

export interface JobStatus {
  id: string;
  job_type: string;
  status: string;
  input_data: Record<string, unknown> | null;
  result: Record<string, unknown> | null;
  error: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface PerformerInfo {
  ticker: string;
  daily_change_pct: number;
}

export interface DashboardSummary {
  best_performer: PerformerInfo | null;
  worst_performer: PerformerInfo | null;
  upcoming_earnings_count: number;
  upcoming_earnings_tickers: string[];
}

File: apps/api/app/schemas/analysis.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class PortfolioSnapshotRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    portfolio_id: int
    created_at: datetime
    total_value: float | None = None
    num_positions: int = 0
    recent_earnings_coverage: float | None = None
    avg_sentiment_score: float | None = None
    risk_exposure_score: float | None = None
    health_score: int | None = None
    concentration_risk: float | None = None
    daily_change: float | None = None
    daily_change_pct: float | None = None


class SectorAllocation(BaseModel):
    sector: str
    weight: float  # 0.0 – 1.0
    value: float


class EarningsInsights(BaseModel):
    holdings_with_recent_earnings: list[str] = []
    positive_outlooks: list[str] = []
    risk_warnings: list[str] = []
    recommended_reviews: list[str] = []
    sentiment_summary: dict[str, int] = {}


class JobStatus(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    job_type: str
    status: str  # pending, processing, completed, failed
    result: dict | None = None
    error: str | None = None
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None


class CompareRequest(BaseModel):
    tickers: list[str] = Field(..., min_length=2, max_length=5)
